interpretation/train_classifier.py

```diff
@@ -0,0 +1,219 @@
+    offset_group = 'well'
+    suffix = '_'.join([norm_antigen, 'norm_per_plate'])
+    pipeline = 'nautilus'
+    suffix = pipeline
+    if norm_antigen is not None:
+        suffix = '_'.join([pipeline, norm_antigen, 'norm_per', norm_group])
+    #%% slice OD table
+    slice_cols = ['pipeline', 'serum ID', 'antigen']
+    slice_keys = [[pipeline], sera_roc_list, ['xkappa-biotin']]
+    slice_actions = ['keep', 'drop', 'drop']
+    for col, action, key in zip(slice_cols, slice_actions, slice_keys):
+        df_norm = slice_df(df_norm, action, col, key)
+    #%% Normalize OD and transform the table to wide format for model training
+    df_norm = normalize_od(df_norm, norm_antigen, norm_group)
+    df_norm = offset_od(df_norm, offset_antigen, offset_group)
+    df_norm['antigen_row'] = df_norm['antigen_row'].map(str)
+    df_norm['antigen_col'] = df_norm['antigen_col'].map(str)
+    pysero_df_pivot = df_norm.copy()
+    pysero_df_pivot = pd.pivot_table(pysero_df_pivot, values='OD',
+                                 index=['plate ID', 'well_id', 'serum ID',
+           'serum type', 'serum dilution', 'secondary ID',
+                        'secondary dilution', 'pipeline'],
+                                 columns=['antigen', 'antigen_row', 'antigen_col'])
+    pysero_df_pivot.columns = ["_".join(cols) for cols in pysero_df_pivot.columns]
+    features = pysero_df_pivot.columns.tolist()
+    pysero_df_pivot.reset_index(inplace=True)
+    # "positive" = 1, "negative" = 0
+    pysero_df_pivot['target'] = (pysero_df_pivot['serum type'] == 'positive')
+    #%% Split the dataset into train and test sets
+    rand_seed = 0
+    gss = GroupShuffleSplit(test_size=.4, n_splits=2, random_state=rand_seed)
+    (train_ids, test_ids), _ = gss.split(pysero_df_pivot, groups=pysero_df_pivot['serum ID'])
+    train = pysero_df_pivot.iloc[train_ids]
+    test = pysero_df_pivot.iloc[test_ids]
+    #%% Initialize xgboost tree classifier
+    clf = XGBClassifier(
+     learning_rate=0.01,
+     n_estimators=1000,
+     max_depth=5,
+     min_child_weight=1,
+     gamma=0,
+     subsample=0.8,
+     colsample_bytree=0.8,
+     objective='binary:logistic',
+     nthread=8,
+     scale_pos_weight=1,
+     seed=0)
+    #%% set up CV folds by serum ID
+    gkf = GroupKFold(n_splits=4)
+    folds = []
+    for fold in gkf.split(train, groups=train['serum ID']):
+        folds.append(fold)
+    #%% Tune xgboost parameter
+    param_tests = [{
+     'max_depth': range(1, 8, 1),
+     'min_child_weight':range(1, 8, 1)
+    },
+    {
+     'gamma':[i/10.0 for i in range(0,5)]
+    },
+    {
+     'subsample':[i/10.0 for i in range(6,10)],
+     'colsample_bytree':[i/10.0 for i in range(6,10)]
+    },
+    {
+     'reg_alpha':[0, 1e-5, 1e-2, 0.1, 1, 100]
+    }
+    ]
+    for param_test in param_tests:
+        clf = tune_cls_para(clf, train, features, target='target', param_test=param_test, cv=folds, n_jobs=-1)
+
+    #%% Train the classifier with optimal parameters
```

Comment:
This comment just confused me because the optimal parameters are from tune_cls_para, or did you find optimal values for learning rate and number of estimators somehow as well?

Reply:
No just lower the learning rate and increase the number of rounds empirically. Clarified this in the comment now.